source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,87 @@
+==============================
+Implementing a DataManipulator
+==============================
+
+If you want to properly implement a ``DataManipulator`` you have to:
+
+1. Implement the ``DataManipulator`` itself
+#. Implement the ``ImmutableDataManipulator``
+
+Upon completion of the first steps you'll have to get this done too:
+
+#. Register the ``Key`` in the ``KeyRegistry``
+#. Implement the ``DataProcessor``
+#. Implement the ``ValueProcessor`` for each value being represented by the ``DataManipulator``
+
+1. Implementing the DataManipulator
+===================================
+
+Implementing the ``DataManipulator`` is done by extending AbstractData in the easiest case. Heres a short example:
+
+.. code-block:: java
+
+  public class SpongeVelocityData extends AbstractSingleData<Vector3d, VelocityData, ImmutableVelocityData> implements VelocityData {
+ ...
+  }
+
+Let's assume we have a single value based manipulator, like for example the ``flying`` state on an entity which can be ``true`` or ``false``.
+We abstracted things above, so now we're able to just extend ``AbstractBooleanData``.
+The next step is to provide the actual ``DataManipulator.class``, the ``Keys.YOUR_MANIPULATOR_NAME_HERE`` and the default value.
+
+.. code-block:: java
+
+ public SpongeVelocityData(Vector3d velocity) {
+          super(VelocityData.class, velocity, Keys.VELOCITY);
+ }
+
+Ok, we're slowly getting there. In most cases while implementing an abstract Manipulator you want to have two constructors:
+
+* One without args (no-args) which calls the second
+* and one single-arg constructor which is actually used for the provided value
+
+Now, after the Manipulator is done you have to check if you are extending ``AbstractData`` or not.
+If you do you must call ``registerFieldGetter()`` and ``registerFieldSetter`` in your constructor.
+Don't forget the ``Field`` registration!
+
+.. code-block:: java
+
+ private void registerStuff() {
+      registerFieldGetter(Keys.MAX_AIR, new GetterFunction<Object>() {
+             @Override
+             public Object get() {
+                 return getMaxAir();
+             }
+         });
+      registerFieldSetter(Keys.MAX_AIR, new SetterFunction<Object>() {
+              @Override
+              public void set(Object value) {
+                  setMaxAir(((Number) value).intValue());
+              }
+          });
+  }
+
+2. ImmutableDataManipulator
+===========================
+
+Repeat the above mentioned steps for the ``ImmutableDataManipulator``.
+If you made it this far, well done.
+
+3. Register the Key in the KeyRegistry
+======================================
+
+The next step is to register your ``KEYS`` in the ``KeyRegistry``.
+
+.. code-block:: java
+
+ public static void registerKeys() {
+      keyMap.put("velocity", makeSingleKey(Vector3d.class, Value.class, of("Velocity")));
+ }
+
+
+4. Implement the DataProcessor
+==============================
+
+Next up is the ``DataProcessor``.
```

Comment:
Brain dump here:

A `DataProcessor` *implements* your `DataManipulator` to work semi-natively with Minecraft's objects. Unfortunately, since it's always easier to just directly implement all of the *processing* of a particular set of data in one place, we have `DataProcessor`s to do just that. 
Currently, a `DataProcessor` has the following methods:
```java
public interface DataProcessor<M extends DataManipulator<M, I>, I extends ImmutableDataManipulator<I, M>> {

    int getPriority();

    boolean supports(DataHolder dataHolder);

    Optional<M> from(DataHolder dataHolder);

    Optional<M> createFrom(DataHolder dataHolder);

    Optional<M> fill(DataHolder dataHolder, M manipulator, MergeFunction overlap);

    Optional<M> fill(DataContainer container, M m);

    DataTransactionResult set(DataHolder dataHolder, M manipulator, MergeFunction function);

    Optional<I> with(Key<? extends BaseValue<?>> key, Object value, I immutable);

    DataTransactionResult remove(DataHolder dataHolder);
}

```
The first thing that we notice is a `getPriority()` method. This is used for multi-processor registration, or in more plain terms: A multi-processor registration system for mods to provide their own compatibility processors for their custom entity/itemstack/tileentity/blockstate types. The higher the priority return, the *earlier* the processor is used in what's called the `DataProcessorDelegate`. The `delegate` is a collection of *registered* `DataProcessor`s of a particular `DataManipulator`such that all registered `DataProcessor`s can be used.

Next up, we have `supports(DataHolder)`. This is really simple, if your `DataManipulator` is supposed to be "supported" by the provided `DataHolder` (which can be an `ItemStack`, `Entity`, `TileEntity`, etc.) you can define this "support" in this method. So, if I have a `VelocityDataProcessor`, my `support(DataHolder)` would look something like so:
```
@Override
public boolean supports(DataHolder dataHolder) {
 return dataHolder instanceof Entity;
}
```

I would finish, but that's enough brain dump for tonight. I'll contribute more brain dump tomorrow.